atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/PerformanceResults.java

```diff
@@ -98,22 +99,65 @@ private static BufferedWriter openFileWriter(File file) throws FileNotFoundExcep
         return new BufferedWriter(new OutputStreamWriter(new FileOutputStream(file), StandardCharsets.UTF_8));
     }
 
-    private static List<Double> getData(RunResult result) {
-        return result.getBenchmarkResults().stream()
-                .flatMap(b -> getRawResults(b.getPrimaryResult().getStatistics()).stream())
-                .collect(Collectors.toList());
+    @VisibleForTesting
+    static List<Double> getData(RunResult result) {
+        return getRawResults(result.getPrimaryResult().getStatistics());
     }
 
     private static List<Double> getRawResults(Statistics statistics) {
+
         try {
             Field field = statistics.getClass().getDeclaredField("values");
             field.setAccessible(true);
             Multiset<Double> rawResults = (Multiset<Double>) field.get(statistics);
-            return rawResults.entrySet().stream()
-                    .flatMap(e -> DoubleStream.iterate(e.getKey(), d -> d).limit(e.getValue()).boxed())
-                    .collect(Collectors.toList());
+            return downSample(rawResults);
         } catch (NoSuchFieldException | IllegalAccessException e) {
-            return Lists.newArrayList();
+            String msg = new StringBuilder().append("Could not get values from statistics !")
+                    .append("\n\t statistics.class = ").append(statistics.getClass().getName())
+                    .append("\n\t statistics.size = ").append(statistics.getN())
+                    .toString();
+            throw new RuntimeException(msg, e);
+        }
+    }
+
+    private static List<Double> downSample(Multiset<Double> multisetParam) {
+
+        final TreeMultiset<Double> values = asTreeMultiset(multisetParam);
+        final long totalCount = values.size();
+
+        if (totalCount <= DOWNSAMPLE_MAXIMUM_SIZE) {
```

Comment:
It was necessary to not upsample when you actually have fewer than DOWNSAMPLE_MAXIMUM_SIZE entries (i.e. 1 entry would be multiplied 500 times otherwise). But the test that I wrote didn't make that clear so I added another one.